File: cpp/1326.minimum-number-of-taps-to-open-to-water-a-garden.cpp
class Solution {
public:
    int minTaps(int n, vector<int>& ranges) {
        vector<int> maxRanges(n + 1, 0);

        for (int i = 0; i < n + 1; i++) {
            int start = max(0, i - ranges[i]);
            int end = min(n, i + ranges[i]);
            maxRanges[start] = max(maxRanges[start], end);
        }

        int taps = 0;
        int maxEnd = 0;
        int curEnd = 0;

        for (int j = 0; j <= n; j++) {
            if (maxEnd < j){
                return -1;
            }

            if (curEnd < j) {
                curEnd = maxEnd;
                taps++;
            }

            if (maxEnd < maxRanges[j]) {
                maxEnd = maxRanges[j];
            }
        }

        return taps;
    }
};

File: golang/1326.minimum-number-of-taps-to-open-to-water-a-garden.go
func minTaps(n int, ranges []int) int { 
    maxRanges := make([]int, n + 1)
    for i := 0; i < n + 1; i++ {
        start, end := max(0, i - ranges[i]), min(n, i + ranges[i])
        maxRanges[start] = max(maxRanges[start], end)
    }

    taps := 0
    maxEnd, curEnd := 0, 0

    for i := 0; i < n + 1; i++ {
        if maxEnd < i {
            return -1
        }

        if curEnd < i {
            curEnd = maxEnd
            taps++
        }

        if maxEnd < maxRanges[i] {
            maxEnd = maxRanges[i]
        }
    }
    return taps
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}